text/0000-target-bundles.md

```diff
@@ -0,0 +1,245 @@
+- Feature Name: target_bundles
+- Start Date: 2016-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Combine distribution of standard libraries and targets into bundles for targeting a particular
+platform. Such bundle is all you need from the Rust side to cross-compile for your target.
+
+The major ideas of this RFC is to:
+
+1. Make JSON targets as full featured as they need to be in order to support specifying targets as
+   custom as they need to be;
+2. Convert current built-in targets to the JSON targets;
+3. Change distribution of libstd to include a corresponding JSON target.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there’s two different ways rustc targets are distributed: built-in targets and custom
+JSON targets. Built-in targets are very inflexible – they cannot be changed without changing and
+recompiling the compiler itself. On the other side of the spectrum, custom JSON targets are easy to
+adjust and adapt, but feature wise are very limited and rarely are suitable for the more uncommon
+use-cases.
+
+We’ve observed a considerable amount of desire by users of the language to customize targets they
```

Comment:
>  it solves uniqueness at one point in time, but if the target spec is edited, you can now have a target dir that contains a mix of objects built under different target spec contents. That's unsafe.

Very true. But using a hash in a directory name as a way of avoiding that unsafety isn't perfect. For avoiding unsafety, we'd really want to embed the target hash in the object files.

> That's really, really nasty if we put the hash of the target in the path to rustlib 

Very true. Which is why the beginning of my comment discards the idea of using a hash.